### PR TITLE
TST: add test to ensure that df.groupby() returns the missing categories when grouping on 2 pd.Categoricals

### LINKED ISSUE #23865: Inconsistant behaviour of empty groups when grouping with one vs. many
Not sure if this is a bug or expected behaviour, but it's something that catches me out constantly. 

```python
import pandas as pd
# Dummy data. Each row is a day and I'm recording where I worked, 
# what percentage of my time was spent coding, 
# and how many cups of coffee I drank
df = pd.DataFrame({
    'place' : ['office','cafe','office','cafe','office','cafe','cafe','cafe','office','office'],
    'percentage':[5, 15, 72,24, 34, 35, 37, 47, 78, 67],
    'coffee': [1,2,0,0,1,2,0,0,1,1] })

# this includes zero counts
df.groupby(pd.cut(df['percentage'], range(0, 100, 10))).size()

# this doesn't 
df.groupby(
    [pd.cut(df['percentage'], range(0, 100, 10)),
    'place']
     ).size()

```
#### Problem description

I want to look at the distribution of my `percentage` columns, so I do `groupby` so that I can use sensible bins. This is my first `gropuby` in the code example. The output is as expected:

```python
percentage
(0, 10]     1
(10, 20]    1
(20, 30]    1
(30, 40]    3
(40, 50]    1
(50, 60]    0
(60, 70]    1
(70, 80]    2
(80, 90]    0
dtype: int64
```
Including the bins for which there were zero observations (50-60% and 80-90%). 

Now I want to also group by place. This is my second `groupby`. Now my empty groups disappear:

```python
percentage  place 
(0, 10]     office    1
(10, 20]    cafe      1
(20, 30]    cafe      1
(30, 40]    cafe      2
                 office    1
(40, 50]    cafe      1
(60, 70]    office    1
(70, 80]    office    2
dtype: int64
```
When I `unstack` this to make a summary table the empty percentage bins are missing:

```python
place       cafe  office
percentage              
(0, 10]      NaN     1.0
(10, 20]     1.0     NaN
(20, 30]     1.0     NaN
(30, 40]     2.0     1.0
(40, 50]     1.0     NaN
(60, 70]     NaN     1.0
(70, 80]     NaN     2.0
```
This will cause a problem when I go to plot this data, as the spacing on my 'percentage' axis will not be consistent. If I'm just counting the size of the bins then I can get round the problem by reindexing with the original categories:

```python
df.groupby(
    [pd.cut(df['percentage'], range(0, 100, 10)),
    'place']
     ).size().unstack().reindex(pd.cut(df['percentage'], range(0, 100, 10)).cat.categories).fillna(0).stack()

percentage  place 
(0, 10]     cafe      0.0
            office    1.0
(10, 20]    cafe      1.0
            office    0.0
(20, 30]    cafe      1.0
            office    0.0
(30, 40]    cafe      2.0
            office    1.0
(40, 50]    cafe      1.0
            office    0.0
(50, 60]    cafe      0.0
            office    0.0
(60, 70]    cafe      0.0
            office    1.0
(70, 80]    cafe      0.0
            office    2.0
(80, 90]    cafe      0.0
            office    0.0
dtype: float64
```
but this seems like a complicated solution for what should be a relatively common problem. And if I want to do some other type of aggregation, e.g. calculate the mean number of cups of coffee for each group, I can't figure it out:

```python
df.groupby(
    [pd.cut(df['percentage'], range(0, 100, 10)),
    'place']
     )['coffee'].mean()

percentage  place 
(0, 10]     office    1.0
(10, 20]    cafe      2.0
(20, 30]    cafe      0.0
(30, 40]    cafe      1.0
            office    1.0
(40, 50]    cafe      0.0
(60, 70]    office    1.0
(70, 80]    office    0.5
```

I have missing groups, but I can't fix it with `stack/fillna/unstack` as I don't want to fill in a value - I want to leave it as missing data, but still  have the group appear. 

Reading through previous issues, it sounds like I am describing this:

https://github.com/pandas-dev/pandas/issues/8138

but the thread says it's fixed, so I can't figure out what is different in my situation. 

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Linux
OS-release: 4.15.0-38-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8
LOCALE: en_GB.UTF-8

pandas: 0.23.4
pytest: None
pip: 10.0.1
setuptools: 40.0.0
Cython: None
numpy: 1.15.1
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.4.0
sphinx: 1.8.1
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### LINKED ISSUE #27075: Groupby ignores unobserved combinations when passing more than one categorical column even if observed=True
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

df_test = pd.DataFrame()
df_test['A'] = pd.Series(np.arange(0,2), dtype='category').cat.set_categories(list(range(0,3)))
df_test['B'] = pd.Series(np.arange(10,12), dtype='category').cat.set_categories(list(range(10,13)))

print("Test DF:")
print(df_test)

print("\nThe following are as expected, unobserved categories have size = 0:")
print(df_test.groupby('A').size())
print(df_test.groupby('B').size())

print("\nThe following does not consider categories, I would expect 9 result lines here:")
print(df_test.groupby(['A','B']).size())

print("\nExpected:")
print(pd.DataFrame({'A':list(range(0,3))*3, 'B':list(range(10,13))*3, '':[1]*2+[0]*7 }).set_index(['A','B']))
```
image

#### Problem description
groupby([cols]) gives back a result for all categories if only one column that is categorical is provided (e.g. ['A']), but it only shows the observed combinations if multiple categorical columns are provided ['A', 'B'], regardless of the setting of observed. I would expect that I would get a result for all combinations of the categorical columns.

#### Expected Output
A result for all combinations of the categorical categories of the groupby columns. For the example above:
A B    
0 10  1
1 11  1
2 12  0
0 10  0
1 11  0
2 12  0
0 10  0
1 11  0
2 12  0

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.3.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.24.2
pytest: 4.5.0
pip: 19.1.1
setuptools: 41.0.1
Cython: 0.29.7
numpy: 1.16.3
scipy: 1.2.1
pyarrow: None
xarray: None
IPython: 7.5.0
sphinx: 2.0.1
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: 1.2.1
tables: 3.5.1
numexpr: 2.6.9
feather: None
matplotlib: 3.0.3
openpyxl: 2.6.2
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.8
lxml.etree: 4.3.3
bs4: 4.7.1
html5lib: 1.0.1
sqlalchemy: 1.3.3
pymysql: None
psycopg2: 2.7.6.1 (dt dec pq3 ext lo64)
jinja2: 2.10.1
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.7.0
gcsfs: None
</details>

### PR DESCRIPTION
**Pull request by smithto1**
- [x] closes #23865
- [x] closes #27075 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

1) _New Tests Addeded_
#23865 #27075 both reported that when df.groupby was called and `by` was set to more than one pd.Categorical column, that any missing categories were not returned, even when observed=False. This issue was fixed in #29690. This Pull Request adds tests to make sure that this correct behaviour is enforced by tests. 

2) _New Bug Found_
Testing did reveal one further issue: DataFrameGroupBy.count() returns NaN for missing Categories, when it should return a count of 0.  SeriesGroupBy.count() does return 0, which is the expected behaviour. I have raised an issue for this bug (#35028 ) and marked the test with an xfail. When the bug is fixed, the xfail will cause the tests to fail, and the xfail can be removed. 

3) _Existing Test Changed as it had the Wrong Expected Result_
A similar issue was reported for .sum() in #31422: missing categories return a sum of NaN when they should return a sum of 0. There was a mistake on the existing test for SeriesGroupBy.sum(), as it said the expected output was NaN (see below) when it should have been 0.
https://github.com/pandas-dev/pandas/blob/0159cba6eb14983ab7eaf38ff138c3c397a6fe3b/pandas/tests/groupby/test_categorical.py#L1315

I have changed this so that the expected output is 0 (this is inline with the comment here: https://github.com/pandas-dev/pandas/issues/31422#issuecomment-579826193 ) and marked the tests for .sum() with xfail. When the bug is addressed, the xfail will cause the tests to fail, and the xfails can be removed.

### CODE REVIEW

File: pandas/tests/groupby/test_categorical.py [hunk 1]
```diff
@@ -1289,12 +1288,12 @@ def test_series_groupby_on_2_categoricals_unobserved(
     assert len(result) == expected_length
 
 
-@pytest.mark.parametrize(
-    "func, zero_or_nan",
+_results_for_groupbys_with_missing_categories = dict(
```
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py:1291]: can you move this to the top & add some comments on the purpose here.

File: pandas/tests/groupby/test_categorical.py [hunk 2]
```diff
@@ -1312,13 +1311,34 @@ def test_series_groupby_on_2_categoricals_unobserved(
...
+def test_series_groupby_on_2_categoricals_unobserved_zeroes_or_nans(
+    reduction_func: str, request
+):
     # GH 17605
     # Tests whether the unobserved categories in the result contain 0 or NaN
+
+    if reduction_func == "ngroup":
+        pytest.skip("ngroup is not truly a reduction")
+
+    if reduction_func == "corrwith":  # GH 32293
```
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py:1329]: make sure the xfail reasons have the issue number as well

File: pandas/tests/groupby/test_categorical.py [hunk 3]
```diff
@@ -1342,6 +1364,83 @@ def test_series_groupby_on_2_categoricals_unobserved_zeroes_or_nans(func, zero_o
...
+
+    res, unobserved_cats = _dataframe_groupby_on_2_categoricals(
+        reduction_func, observed=True
+    )
+
+    for cat in unobserved_cats:
+        assert cat not in res.index
+
+
+def _dataframe_groupby_on_2_categoricals(reduction_func: str, observed: bool):
```
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py:1383]: can you add some commentary here. i don't normally like having a helper function like this as hard for folks to grok, but i see might be hard to remove

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_categorical.py (hunk 1 above)
Comment (smithto1) [reply pandas/tests/groupby/test_categorical.py]: Done

File: pandas/tests/groupby/test_categorical.py (hunk 2 above)
Comment (smithto1) [reply pandas/tests/groupby/test_categorical.py:1337]: Done

File: pandas/tests/groupby/test_categorical.py (hunk 3 above)
Comment (smithto1) [reply pandas/tests/groupby/test_categorical.py]: Just removed the helper function, not hard to do. Just requires repeating some of the same code across two tests.

**Comment by smithto1** [comment]
@jreback Can I get a another review/approval on this? I'd like to close it off and continue addressing the related bugs. 

Addressed all your previous comments and merged in the latest master with all checks passing.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @smithto1 very nice, very happy to have follows to fix things :->

**Pull request merged**